### PR TITLE
Add disable external url option

### LINKED ISSUE #96: Need an option to diable externalURL
### What feature do you want to see added?

In some closed networks, users are unable to view artifact files using Azure Storage's Bob SAS URL. 
An option is needed to disable the feature.

### Upstream changes

_No response_

### Are you interested in contributing this feature?

yes

### PR DESCRIPTION
**Pull request by WenjunMao**
Fixes #96
Add disable external url option. 
One option can disable Jenkins from generating SAS URLs when users want to view artifact files in a closed network.

### CODE REVIEW

File: src/main/resources/com/microsoft/jenkins/artifactmanager/AzureArtifactConfig/help-disableExternalUrl.html [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<p>
+    if checked, Azure Storage will not generate a permalink via a Blob SAS URL for the file.
```
Comment (timja) [inline src/main/resources/com/microsoft/jenkins/artifactmanager/AzureArtifactConfig/help-disableExternalUrl.html:2]:
```suggestion
    If checked, artifacts will be accessed under a Jenkins URL and not an Azure Storage URL.
```

File: src/main/java/com/microsoft/jenkins/artifactmanager/AzureBlobVirtualFile.java [hunk 2]
```diff
@@ -185,20 +188,24 @@ public URI toURI() {
     @CheckForNull
     @Override
     public URL toExternalURL() throws IOException {
-        StorageAccountInfo accountInfo = Utils.getStorageAccount(build.getParent());
-        String sas;
-        try {
-            sas = Utils.generateBlobSASURL(accountInfo, this.container, this.key);
-        } catch (Exception e) {
-            throw new IOException(e);
+        if (this.disableExternalUrl == null || !this.disableExternalUrl) {
```
Comment (timja) [inline src/main/java/com/microsoft/jenkins/artifactmanager/AzureBlobVirtualFile.java:191]:
```suggestion
        if (!this.disableExternalUrl) {
```

File: src/main/java/com/microsoft/jenkins/artifactmanager/AzureBlobVirtualFile.java [hunk 3]
```diff
@@ -49,12 +49,15 @@ public class AzureBlobVirtualFile extends AzureAbstractVirtualFile {
 
     private final String container;
     private final String key;
+
+    private Boolean disableExternalUrl;
     private final transient Run<?, ?> build;
 
-    public AzureBlobVirtualFile(String container, String key, Run<?, ?> build) {
+    public AzureBlobVirtualFile(String container, String key, Boolean disableExternalUrl, Run<?, ?> build) {
```
Comment (timja) [inline src/main/java/com/microsoft/jenkins/artifactmanager/AzureBlobVirtualFile.java:56]:
```suggestion
    public AzureBlobVirtualFile(String container, String key, boolean disableExternalUrl, Run<?, ?> build) {
```

File: src/main/java/com/microsoft/jenkins/artifactmanager/AzureArtifactConfig.java [hunk 4]
```diff
@@ -61,6 +62,15 @@ public void setPrefix(String prefix) {
         this.prefix = prefix;
     }
 
+    public Boolean getDisableExternalUrl() {
+        return disableExternalUrl;
+    }
+
+    @DataBoundSetter
+    public void setDisableExternalUrl(Boolean disableExternalUrl) {
```
Comment (timja) [inline src/main/java/com/microsoft/jenkins/artifactmanager/AzureArtifactConfig.java:70]:
```suggestion
    public void setDisableExternalUrl(boolean disableExternalUrl) {
```

File: src/main/java/com/microsoft/jenkins/artifactmanager/AzureArtifactConfig.java [hunk 5]
```diff
@@ -61,6 +62,15 @@ public void setPrefix(String prefix) {
         this.prefix = prefix;
     }
 
+    public Boolean getDisableExternalUrl() {
```
Comment (timja) [inline src/main/java/com/microsoft/jenkins/artifactmanager/AzureArtifactConfig.java:65]:
```suggestion
    public boolean getDisableExternalUrl() {
```

File: src/main/java/com/microsoft/jenkins/artifactmanager/AzureArtifactConfig.java [hunk 6]
```diff
@@ -34,6 +34,7 @@ public class AzureArtifactConfig extends AbstractDescribableImpl<AzureArtifactCo
     private String storageCredentialId;
     private String container;
     private String prefix;
+    private Boolean disableExternalUrl;
```
Comment (timja) [inline src/main/java/com/microsoft/jenkins/artifactmanager/AzureArtifactConfig.java:37]:
```suggestion
    private boolean disableExternalUrl;
```

**Comment by WenjunMao** [comment]
> > If the external URL can be disabled, using a private endpoint would be a good option.
> 
> You don't need to disable the external URL then, if you setup the private endpoint and the DNS resolution, Azure will use its 'magic' DNS so that the public url will resolve a CNAME for the private endpoint and your users will hit the private endpoint NIC and not the public address

In my special network, it is very difficult to use that magic.

**Review by timja — APPROVED** [review]

**Pull request merged**